pennylane/qnode.py

```diff
@@ -838,8 +856,13 @@ def to_tf(self, dtype=None):
 
             if self.interface != "tf" and self.interface is not None:
                 # Since the interface is changing, need to re-validate the tape class.
+                # if method was changed from "best", set it back to best
+                if self.diff_method_change:
+                    diff_method = "best"
+                else:
+                    diff_method = self.diff_method

```

Comment:
Setting `diff_method` to "best" would potentially change `self.diff_options["method"]` to something else. In that case `self.diff_method` would still have the former changed value. I think that the following part might need to be added for all the interface changes, after `self.get_tape`, as well, right?

```python
# Update diff_method to reflect which method has been selected
if self.diff_method == "best":
    self.diff_method_change = True
    if tape_diff_options["method"] == "device":
        self.diff_method = "device"
    elif tape_diff_options["method"] == "backprop":
        self.diff_method = "backprop"
    elif tape_diff_options["method"] == "best":
        self.diff_method = "parameter-shift"
    elif tape_diff_options["method"] == "numeric":
        self.diff_method = "finite-diff"
```

I'd suggest, in that case, to put this in a separate method and call it from both the `__init__` method as well as the interface-changing methods (`to_autograd`, etc.). This could then also be tested in a better way by asserting that this new method has been called when using the "best" differentiation method.

Sorry for discovering this so late, and thanks for keeping up the good work! 🙂

Reply:
Yes! That's a good catch. Once we set 'best' again here, we are not changing it later. This should be done to give the correct `diff_method` after interface change.

I agree with this. No point in copy pasting the same lines again and again.
>  put this in a separate method and call it from both the __init__ method as well as the interface-changing methods (to_autograd, etc.).

I'll start on this.

Reply:
This is a good catch. It might happen that after an interface change, the `self.diff_options["method"]` has been updated. In this case, `diff_method` will still show the old value.
What you suggest is a simple enough fix.
I have implemented it.
Now I'll see what can be done with tests.

Reply:
@thisac I don't know how to assert for method call. Can you give me some pointers?

Reply:
Thanks @ankit27kh! You can read more about `mocker.spy` [here](https://pypi.org/project/pytest-mock/), but basically you create a 

```python
spy = mocker.spy(object, "method")
``` 

and then make sure it's called once with 

```python
spy.assert_called_once()
```

There are a few cases in the `test_qnode.py` file as well (e.g., [here](https://github.com/PennyLaneAI/pennylane/blob/f147f22d8d99ba5891edd45a6a1f7dd679c8a23c/tests/tape/test_qnode.py#L430)).

Reply:
Thanks!
I have added this.